exercises/Mathieu_Tillet/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,125 @@
+        out_name = "./out.txt";
+    }
+
+    ifstream in;
+    in.open(file_name);
+
+    if (!in){
+        cerr << "Error when reading the file" << endl;
+        return -1;
+    }
+
+    string block;
+    size_t max_length = 4; // size of "Word", which we systematically write
+    size_t block_size;
+    bool should_split;
+
+    // the delimiters of words
+    string delimiters = ".,?'\"!():";
+
+    while (in >> block){
+        // iterate on each space or endofline separated block in the text
+        block_size = block.size();
+        size_t start = 0;
+        should_split = false;
+
+        // iterate on each character of the block to further parse it
+        for (size_t i = 0; i != block_size; ++i){
+            // convert to lowercase so it is case insensitive
+            block[i] = tolower(block[i]);
+
+            // iterate on every delimiter to check if the current character is a letter
+            for (size_t j = 0; j != delimiters.size(); j++){
+                if (delimiters[j] == block[i]){
+                    should_split = true;
+                    break;
+                }
+            }
+            // if the last character was invalid, we reached the end of a word
+            if (should_split){
+                size_t length_of_word = i - start;
+                if (length_of_word >= 5){
+                    // if long enough, the word is added to the map
+                    word_count[block.substr(start, length_of_word)]++;
+
+                    // check if it is the new longest word
+                    if (length_of_word > max_length){
+                        max_length = length_of_word;
+                    }
+                }
+                should_split = false;
+                start = i + 1;
+            }
+        }
+        // if the last character was valid, the last part of the block is checked for adding
+        if (!should_split){
+            if (block_size - start >= 5){
+                word_count[block.substr(start, block_size - start)]++;
+                if (block_size - start > max_length){
+                    max_length = block_size - start;
+                }
+            }
+        }
+    }
+    in.close();
+
+    // we store the elements of word_counts in a vector so they can be sorted
+    vector<tuple<int, string>> vec_count;
+    for (map<string, int>::iterator it = word_count.begin(); it != word_count.end(); ++it){
+        vec_count.push_back(tuple<int, string>(it->second, it->first));
+    }
+
+    // sort the vector's elements in increasing order, by the first value of the tuple (ie the number of occurences)
+    sort(vec_count.begin(), vec_count.end());
+
+    ofstream out;
+    size_t spaces;
+    out.open(out_name);
+    out << "Word     Usage" << endl;
+    // we go through the elements of the vector in reverse order so that they are sorted decreasingly in the output
+    for (vector<tuple<int, string>>::iterator it = vec_count.end(); it != vec_count.begin(); --it){
```

Comment:
You should be able to use a range based loop here `for (const auto& item : vector_count)`